sql-plugin/src/main/scala/com/nvidia/spark/rapids/RmmRapidsRetryIterator.scala

```diff
@@ -0,0 +1,360 @@
+      new RmmRapidsRetryAutoCloseableIterator(Seq(wrapped).iterator, fn))
+  }
+
+  /**
+   * Helper method to drain an iterator and ensuring that it was non-empty
+   * and it had a single item in it.
+   */
+  private def drainSingleWithVerification[K](it: Iterator[K]): K = {
+    require(it.hasNext,
+      "Couldn't drain a single item with a closed iterator!")
+    val item = it.next()
+    require(!it.hasNext,
+      "Multiple items found in the source iterator but one expected!")
+    item
+  }
+
+  /**
+   * AutoCloseable wrapper on Seq[T], returning a Seq[T] that can be closed.
+   * @param ts the Seq to wrap
+   * @tparam T the type of the items in `ts`
+   */
+  private case class AutoCloseableSeqInternal[T <: AutoCloseable](ts: Seq[T])
+      extends Seq[T] with AutoCloseable{
+    override def close(): Unit = {
+      ts.foreach(_.safeClose())
+    }
+
+    override def length: Int = ts.length
+
+    override def iterator: Iterator[T] = ts.iterator
+
+    override def apply(idx: Int): T = ts.apply(idx)
+  }
+
+  /**
+   * RmmRapidsRetryAutoCloseableIterator exposes an iterator that can retry work,
+   * specified by `fn`, abstracting away the retry specifics. Elements passed to this iterator
+   * must be AutoCloseable.
+   *
+   * It assumes the type K is AutoCloseable, and that if a split policy is specified, that it
+   * is capable of handling splitting one K into a sequence of them.
+   *
+   * When an attempt to invoke function `fn` is successful, the item K in `input` will be
+   * closed. In the case of a failure, all items in `input` will be closed.
+   *
+   * @tparam K element type that must be AutoCloseable
+   * @tparam T `fn` result type
+   * @param input an iterator of K
+   * @param fn a function that takes K and produces T
+   * @param splitPolicy a function that can split an item of type T into a Seq[T]. The split
+   *                    function must close the item passed to it.
+   */
+  class RmmRapidsRetryAutoCloseableIterator[K <: AutoCloseable, T](
+      input: Iterator[K],
+      fn: K => T,
+      splitPolicy: K => Seq[K])
+      extends RmmRapidsRetryIterator(input, fn, splitPolicy)
+        with Arm {
+
+    def this(
+        input: Iterator[K],
+        fn: K => T) = {
+      this(input, fn, null)
+    }
+
+    override def invokeFn(k: K): T = {
+      val res = super.invokeFn(k)
+      k.close() // close `k` only if we didn't throw from `invokeFn`
+      res
+    }
+
+    override def hasNext: Boolean = super.hasNext
+
+    override def next(): T = {
+      if (!hasNext) {
+        throw new IllegalStateException("Closed called on an empty iterator.")
+      }
+      try {
+        super.next()
+      } catch {
+        case t: Throwable =>
+          // exception occurred while trying to handle this retry
+          // we close our inputs, and attempts (which includes the item we last attempted)
+          (input ++ attemptStack.iterator).toArray[AutoCloseable].safeClose(t)
```

Comment:
I believe also closing the `input` is a bad idea here, especially if the iterator behind `input` is really expensive to manifest. I believe we want callers (e.g `withRetry` code) to handle the closing of inputs, whereas the `attemptStack` can be handled here.

Reply:
handled here https://github.com/NVIDIA/spark-rapids/pull/7771/commits/08f6d8f37e2707c7fbbd8ea34fc51d17ae92726e